src/number/streaming.rs

```diff
@@ -1578,6 +1663,79 @@ where
   Ok((i, float))
 }
 
+/// Recognizes floating point number in text format and returns a f64.
+/// This only handles non-finite (special) values.
+pub fn double_nonfinite<T, E: ParseError<T>>(input: T) -> IResult<T, f64, E>
+where
+  T: Slice<RangeFrom<usize>> + Slice<RangeTo<usize>>,
+  T: Clone + Offset,
+  T: InputIter + InputLength + InputTake + crate::traits::ParseTo<i32>,
+  <T as InputIter>::Item: AsChar,
+  <T as InputIter>::IterElem: Clone,
+  T: InputTakeAtPosition,
+  <T as InputTakeAtPosition>::Item: AsChar,
+  T: AsBytes,
+  T: for<'a> Compare<&'a [u8]>,
+{
+  let (i, sign) = sign(input.clone())?;
```

Comment:
the sign parsing is done twice in case double_finite fail.

Reply:
True, I can move that out. Good suggestion.

Reply:
Wait, the one issue with that is our error handling isn't identical in these two cases. The double sign check might be optimized out by the compiler.

If we return `Err(Err::Error(E::from_error_kind(input, ErrorKind::Float)));` relative to the sign, and not relative to the start of the buffer, we get a different result. If the float format is not valid, then I believe the returned input should start before the sign, rather than afterwards. We could also pass both `input` and the relative input to the function, but that's a bit of API clutter.

Generally, special floats tend to be quite rare in real datasets, so performance for the specials is less of a necessity than other cases.

Reply:
AFAIK to nom convention it's totally ok to say the problem is after the sign cause... it is. I think... anyway the only thing I know to debug a nom parser is to use [nom_tracable](https://crates.io/crates/nom-tracable)

Reply:
This has been implemented, the general approach now parses the sign once, and then only handles adding the right sign after all other parsing has been done:

https://github.com/Alexhuszagh/nom/blob/805e5be9d7e16a6fc0d86e3e6a16fd042d97fd3f/src/number/streaming.rs#L1597-L1620